### PR TITLE
`Programming exercises`: Fix an issue with access tokens for team exercises

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

For programming team exercises, the vcs access token did not load, as the team's students were not loaded when fetching the participation.

### Description

Fix the issue by loading the team's students, when the exercise is a team exercise.

### Steps for Testing

Prerequisites:
1. As an instructor create a team programming exercise
2. Create a team and add at least one student
3. With the student account, login and navigate to the exercise
4. Make sure the clone link in the code button is displayed with the VCS-access token.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| ParticipationVcsAccessTokenService.java | 92% | ✅ ❌ |

### Screenshots

image
image

## Release Notes

- **New Features**
  - Introduced functionality to manage team exercises, including loading team student data and creating access tokens for team participation.
  - Added methods to handle programming submissions specifically for team exercises.
  - Implemented a method to delete specific teams from the repository.

- **Bug Fixes**
  - Enhanced logic for retrieving and creating VCS access tokens to ensure team data is accurately processed.

- **Tests**
  - Added new test methods to verify the functionality related to team exercises and VCS access tokens.

### CODE REVIEW

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Token appeared as described

image

**Review by EneaGore — APPROVED** [review]
Tested on TS3. works as expected. code also lgtm

**Review by alekspetrov9e — APPROVED** [review]
Tested on TS1. Works fine.

**Review by az108 — CHANGES_REQUESTED** [review]
Overall Code LGTM, only 2 suggestions

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ParticipationVcsAccessTokenService.java [hunk 1]
```diff
@@ -81,6 +87,17 @@ public ParticipationVCSAccessToken createVcsAccessTokenForUserAndParticipationId
...
     }
 
+    /**
+     * Loads the team students of a participation's team, if it has a team
+     *
+     * @param participation the participation which team's students are not loaded yet
+     */
+    private void loadTeamStudentsForTeamExercise(StudentParticipation participation) {
+        if (participation.getTeam().isPresent()) {
+            participation.getTeam().get().setStudents(teamRepository.findWithStudentsById(participation.getTeam().get().getId()).get().getStudents());
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ParticipationVcsAccessTokenService.java]: Yeah, I split it for readability

File: src/test/java/de/tum/cit/aet/artemis/exercise/team/TeamUtilService.java [hunk 2]
```diff
@@ -244,4 +244,11 @@ public Team createTeam(Set<User> students, User owner, Exercise exercise, String
         team.setExercise(exercise);
         return teamRepo.saveAndFlush(team);
     }
+
+    /**
+     * Deletes a team
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/test/java/de/tum/cit/aet/artemis/exercise/team/TeamUtilService.java]: Yep, I removed it

**Review by kevinfischer4 — APPROVED** [review]
Tested on TS3, works as described!
image

**Review by az108 — APPROVED** [review]
Thanks for incorporating the changes! Code LGTM 👍

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ParticipationVcsAccessTokenService.java [hunk 3]
```diff
@@ -73,6 +79,7 @@ public ParticipationVCSAccessToken findByUserAndParticipationIdOrElseThrow(User
     public ParticipationVCSAccessToken createVcsAccessTokenForUserAndParticipationIdOrElseThrow(User user, long participationId) {
         participationVcsAccessTokenRepository.findByUserIdAndParticipationIdAndThrowIfExists(user.getId(), participationId);
         var participation = programmingExerciseStudentParticipationRepository.findByIdElseThrow(participationId);
+        loadTeamStudentsForTeamExercise(participation);
```
Comment (BBesrour) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ParticipationVcsAccessTokenService.java:82]: Can't we load the team/students with the previous call?

**Review by BBesrour — APPROVED** [review]
code lgtm, small question that could help reduce the db queries

**Review by ItsaaaMeMario — APPROVED** [review]
Works as expected!

Bildschirmfoto 2024-11-20 um 13 56 02

**Review by Cathy0123456789 — APPROVED** [review]
Tested on TS3. Works as described 
image

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Reapprove

**Review by flbrgit — APPROVED** [review]
Tested on TS3. Everything works as expected
Screenshot 2024-11-23 231148

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ParticipationVcsAccessTokenService.java (hunk 3 above)
Comment (krusche) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ParticipationVcsAccessTokenService.java:82]: I think it's fine to load them conditionally, i.e. only when it is a team exercise

**Review by krusche — APPROVED** [review]

**Pull request merged**